Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,2354 @@
+// TODO rename
+using IntegratorTolerances = CurveSegment::IntegratorTolerances;
+using ObstacleIndex        = CableSpan::ObstacleIndex;
+
+//==============================================================================
+//                            Frenet Frame Helpers
+//==============================================================================
+
+namespace
+{
+
+// Define the frenet frame as the position and orientation with:
+// - position on the geodesic,
+// - tangent along X direction,
+// - surface normal along Y direction,
+// - binormal along Z direction.
+using FrenetFrame                        = Transform;
+static const CoordinateAxis TangentAxis  = CoordinateAxis::XCoordinateAxis();
+static const CoordinateAxis NormalAxis   = CoordinateAxis::YCoordinateAxis();
+static const CoordinateAxis BinormalAxis = CoordinateAxis::ZCoordinateAxis();
+
+const UnitVec3& getAxis(const FrenetFrame& X, CoordinateAxis axis)
+{
+    return X.R().getAxisUnitVec(axis);
+}
+
+const UnitVec3& getTangent(const FrenetFrame& X)
+{
+    return getAxis(X, TangentAxis);
+}
+
+const UnitVec3& getNormal(const FrenetFrame& X)
+{
+    return getAxis(X, NormalAxis);
+}
+
+const UnitVec3& getBinormal(const FrenetFrame& X)
+{
+    return getAxis(X, BinormalAxis);
+}
+
+} // namespace
+
+//==============================================================================
+//              Contact Geometry Helpers For Sign Inversion
+//==============================================================================
+
+// This section contains helper functions for flipping the sign of the output
+// obtained from ContactGeometry.
+namespace
+{
+
+// Flip sign of curvature, such that tangentDot = curvature * normal
+Real calcSurfaceCurvature(
+    const ContactGeometry& geometry,
+    const FrenetFrame& X_S,
+    CoordinateAxis direction)
+{
+    return -geometry.calcSurfaceCurvatureInDirection(
+        X_S.p(),
+        getAxis(X_S, direction));
+}
+
+// Helper function for flipping the sign such that positive value is outside of
+// the surface.
+Real calcSurfaceValue(const ContactGeometry& geometry, const Vec3& point_S)
+{
+    return -geometry.calcSurfaceValue(point_S);
+}
+// Flipping sign...
+Vec3 calcSurfaceGradient(const ContactGeometry& geometry, const Vec3& point_S)
+{
+    return -geometry.calcSurfaceGradient(point_S);
+}
+// Flipping sign...
+Mat33 calcSurfaceHessian(const ContactGeometry& geometry, const Vec3& point_S)
+{
+    return -geometry.calcSurfaceHessian(point_S);
+}
+
+} // namespace
+
+//==============================================================================
+//                          CABLE SUBSYSTEM
+//==============================================================================
+
+bool CableSubsystem::isInstanceOf(const Subsystem& s)
+{
+    return Impl::isA(s.getSubsystemGuts());
+}
+
+const CableSubsystem& CableSubsystem::downcast(const Subsystem& s)
+{
+    assert(isInstanceOf(s));
+    return static_cast<const CableSubsystem&>(s);
```

Comment:
`dynamic_cast`ing the reference would give you the same result *and* assert that it's an instance of the given class

Reply:
`Subsystem` is not polymorphic, hence the `static_cast`